File: Iteration-2/src/CourseRegistrationSystem.java

import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.Iterator;

import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

public class CourseRegistrationSystem {
    private List<Student> students;
    private List<Course> courses;
    private List<Advisor> advisors;
    private List<Transcript> transcripts;

    public List<Student> getStudents() {
        return students;
    }

    public List<Course> getCourses() {
        return courses;
    }
    
    public CourseRegistrationSystem() {
        this.students = new ArrayList<>();
        this.courses = new ArrayList<>();
        this.advisors = new ArrayList<>();
        this.transcripts = new ArrayList<>();
    }

    public void loadStudentsFromJson(String filePath) {
        JSONParser parser = new JSONParser();

        try (FileReader reader = new FileReader(filePath)) {
            Object obj = parser.parse(reader);
            JSONArray studentArray = (JSONArray) obj;

            for (Object studentObj : studentArray) {
                JSONObject studentInFile = (JSONObject) studentObj;
                Student student = new Student(
                        (String) studentInFile.get("firstName"),
                        (String) studentInFile.get("lastName"),
                        (String) studentInFile.get("studentID"),
                        (String) studentInFile.get("password"),
                        ((Double) studentInFile.get("GPA")).doubleValue(),
                        ((Long) studentInFile.get("semester")).intValue()
                );
                students.add(student);
            }
        } catch (IOException | ParseException e) {
            e.printStackTrace();
        }
    }

    public void loadCoursesFromJson(String filePath) {
        JSONParser parser = new JSONParser();

        try (FileReader reader = new FileReader(filePath)) {
            Object obj = parser.parse(reader);
            JSONArray courseArray = (JSONArray) obj;

            for (Object courseObj : courseArray) {
                JSONObject courseInFile = (JSONObject) courseObj;
                Course course = new Course(
                        (String) courseInFile.get("courseID"),
                        (String) courseInFile.get("courseName"),
                        ((Long) courseInFile.get("credit")).intValue(),
                        null,
                        null,
                        (String) courseInFile.get("prerequisite"),
                        ((Long) courseInFile.get("coursesemester")).intValue()
                );
                courses.add(course);
            }
        } catch (IOException | ParseException e) {
            e.printStackTrace();
        }
    }

    public void loadAdvisorsFromJson(String filePath) {
        JSONParser parser = new JSONParser();

        try (FileReader reader = new FileReader(filePath)) {
            Object obj = parser.parse(reader);
            JSONArray advisorArray = (JSONArray) obj;

            for (Object advisorObj : advisorArray) {
                JSONObject advisorInFile = (JSONObject) advisorObj;
                String firstName = (String) advisorInFile.get("firstName");
                String lastName = (String) advisorInFile.get("lastName");
                String password = (String) advisorInFile.get("password");

                JSONArray studentsArray = (JSONArray) advisorInFile.get("students");
                ArrayList<Student> students = new ArrayList<>();

                for (Object studentObj : studentsArray) {
                    JSONObject studentInFile = (JSONObject) studentObj;
                    Student student = new Student(
                            (String) studentInFile.get("firstName"),
                            (String) studentInFile.get("lastName"),
                            (String) studentInFile.get("studentID"),
                            (String) studentInFile.get("password"),
                            ((Double) studentInFile.get("GPA")).doubleValue(),
                            ((Long) studentInFile.get("semester")).intValue()
                    );
                    students.add(student);
                }

                Advisor advisor = new Advisor(firstName, lastName, password, new ArrayList<>(), students);
                advisors.add(advisor);
            }
        } catch (IOException | ParseException e) {
            e.printStackTrace();
        }
    }

    public void authenticateAndEnroll(Person person) {
        Scanner scanner = new Scanner(System.in);
        boolean authenticated = false;

        while (!authenticated) {
            System.out.print("Email address: ");
            String enteredEmail = scanner.nextLine().toLowerCase();
            System.out.print("Password: ");
            String enteredPassword = scanner.nextLine();
          
            if (person instanceof Student) {
                Student currentStudent = (Student) person;
                Transcript studentTranscript = new Transcript(currentStudent.getStudentID());
                String signedEmail = (currentStudent.createEmailAddress()).toLowerCase();
                String signedPassword = currentStudent.getPassword();
                if (signedEmail.equals(enteredEmail) && signedPassword.equals(enteredPassword)) {
                    System.out.println("Student Login successful.");
                    selectCourses(currentStudent,studentTranscript);
                    generateStudentFiles(currentStudent);
                    authenticated = true;
                }
            } else if (person instanceof Advisor) {
                Advisor currentAdvisor = (Advisor) person;
                String signedEmail = (currentAdvisor.createEmailAddress()).toLowerCase();
                String signedPassword = currentAdvisor.getPassword();
                if (signedEmail.equals(enteredEmail) && signedPassword.equals(enteredPassword)) {
                    System.out.println("Advisor Login successful.");
                    displayEnrolledStudentsForAdvisor(currentAdvisor);
                    authenticated = true;
                }
            }

            if (!authenticated) {
                System.out.println("Login failed. Please try again.");
            }
        }
    }
    public boolean hasPrerequisites(Student student, Course course) {
        String prerequisiteCourseID = course.getPrerequisite();

        if (prerequisiteCourseID.isEmpty()) {
            return true;
        }

        for (Course completedCourse : student.getCourses()) {
            if (completedCourse.getCourseID().equals(prerequisiteCourseID)) {
                return true;
            }
        }

        return false;
    }


    void selectCourses(Student authenticatedStudent, Transcript transcript) {
        Scanner scanner = new Scanner(System.in);

        System.out.println("Welcome " + authenticatedStudent.getFirstName() + " " + authenticatedStudent.getLastName() + "\n");

        List<Course> availableCourses = getAvailableCourses(authenticatedStudent);

        if (availableCourses.size() == 0) {
            System.out.println("There are no available courses for your semester.");
            return;
        }

        int coursesEnrolled = 0;
        
        if(hasEnrollmentFile(authenticatedStudent)) {
            String studentFileName2 = authenticatedStudent.getStudentID() + ".json";

            File studentFile2 = new File(studentFileName2);

            if (studentFile2.exists()) {
                JSONParser parser = new JSONParser();

                try (FileReader reader = new FileReader(studentFileName2)) {
                    Object obj = parser.parse(reader);
                    JSONObject studentInFile = (JSONObject) obj;

                    JSONArray enrolledCoursesArray = (JSONArray) studentInFile.get("enrolledCourses");

                    coursesEnrolled = enrolledCoursesArray.size(); 
                    System.out.println("You have succesfully selected " + coursesEnrolled + " courses! ");
                } catch (IOException | ParseException e) {
                    e.printStackTrace();
                }
            } else {
                coursesEnrolled = 0;
            }
        }
        	while (coursesEnrolled < 5) {
                if (availableCourses.size() == 0) {
                    System.out.println("There are no available courses.");
                    return;
                }
                System.out.println("Available Courses:");

                for (int i = 0; i < availableCourses.size(); i++) {
                    Course course = availableCourses.get(i);
                    System.out.println((i + 1) + ". " + course.getCourseName());
                }

                System.out.println("Select courses by entering their numbers one by one, or enter 0 to exit: ");
                String userInput = scanner.nextLine();

                if (userInput.equals("0")) {
                    break;
                }
                
                String[] selectedCourseNumbers = userInput.split(",");
               
                for (String courseNumber : selectedCourseNumbers) {
                    int choice = Integer.parseInt(courseNumber);

                    if (choice >= 1 && choice <= availableCourses.size()) {
                        Course selectedCourse = availableCourses.get(choice - 1);

                        if (hasPrerequisites(authenticatedStudent, selectedCourse)) {

                            authenticatedStudent.addCourse(selectedCourse);
                            System.out.println("Enrolled in course: " + selectedCourse.getCourseID() + " - " + selectedCourse.getCourseName());
                            availableCourses.remove(selectedCourse);
                            coursesEnrolled++;
                        } else {
                            System.out.println("You haven't completed the prerequisite for course: " + selectedCourse.getCourseID());
                        }
                    } else {
                        System.out.println("Inappropriate selection: " + courseNumber + ". Please enter an appropriate course.");
                    }
                }

                System.out.println();
            }
        }
     List<Course> getAvailableCourses(Student authenticatedStudent) {
    	
        List<Course> availableCourses = new ArrayList<>();

        for (Course course : courses) {
            if (course.getCoursesemester() == authenticatedStudent.getSemester() && !authenticatedStudent.getCourses().contains(course)) {
                availableCourses.add(course);
            }
        }

        return availableCourses;
    }

    void generateStudentFiles(Student student) {
            JSONObject studentInFile = new JSONObject();
            JSONArray enrolledCoursesArray = new JSONArray();
            studentInFile.put("studentID", student.getStudentID());
            studentInFile.put("firstName", student.getFirstName());
            studentInFile.put("lastName", student.getLastName());
            studentInFile.put("semester", student.getSemester());
            studentInFile.put("GPA", student.getGPA());
            for (Course course : student.getCourses()) {
                JSONObject courseInFile = new JSONObject();
                courseInFile.put("courseID", course.getCourseID());
                courseInFile.put("courseName", course.getCourseName());
                courseInFile.put("credit", course.getCredit());
                courseInFile.put("prerequisite", course.getPrerequisite());
                courseInFile.put("coursesemester", course.getCoursesemester());
                enrolledCoursesArray.add(courseInFile);
            }

            studentInFile.put("enrolledCourses", enrolledCoursesArray);

            try (FileWriter file = new FileWriter(student.getStudentID() + ".json")) {
                file.write(studentInFile.toJSONString());

            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    

    public void displayEnrolledStudentsForAdvisor(Advisor advisor) {
        List<Student> enrolledStudents = advisor.getStudents();

        if (enrolledStudents.isEmpty()) {
            System.out.println("No students enrolled in courses.");
        } else {
            System.out.println("Enrolled Students:");

            for (Student student : enrolledStudents) {
                if (hasEnrollmentFile(student)) {
                    System.out.println("Student: " + student.getFirstName() + " " + student.getLastName());

                    String studentFileName = student.getStudentID() + ".json";

                    File studentFile = new File(studentFileName);

                    if (studentFile.exists()) {
                        JSONParser parser = new JSONParser();

                        try (FileReader reader = new FileReader(studentFileName)) {
                            Object obj = parser.parse(reader);
                            JSONObject studentInFile = (JSONObject) obj;

                            JSONArray enrolledCoursesArray = (JSONArray) studentInFile.get("enrolledCourses");

                            if (enrolledCoursesArray.isEmpty()) {
                                System.out.println("No enrolled courses for this student.");
                            } else {
                                System.out.println("Enrolled Courses:");
                                
                                
                                
                                
                                Iterator<Object> iterator = enrolledCoursesArray.iterator();
                                while (iterator.hasNext()) {
                                    JSONObject courseInFile = (JSONObject) iterator.next();
                                    String courseID = (String) courseInFile.get("courseID");
                                    String courseName = (String) courseInFile.get("courseName");
                                    System.out.println(courseID + " - " + courseName);

                                    Scanner scanner = new Scanner(System.in);
                                    System.out.println("Do you want to approve (A) or reject (R) the course selection for this student?");
                                    String approvalDecision = scanner.next();

                                    if (approvalDecision.equalsIgnoreCase("A")) {
                                        System.out.println("Course selection approved!");
                                    } else if (approvalDecision.equalsIgnoreCase("R")) {
                                        String courseToRemove1 = courseName;
                                        removeCourseInfoFromStudentFile(studentFileName, courseToRemove1);

                                        iterator.remove();

                                        System.out.println("Course selection rejected!");
                                    } else {
                                        System.out.println("Invalid input. Please enter 'A' for approve or 'R' for reject.");
                                    }
                                }
                            }
                        } catch (IOException | ParseException e) {
                            e.printStackTrace();
                        }
                    } else {
                        System.out.println("No student has chosen courses yet.");
                    }

                    System.out.println("------");
                }
            }
        }
    }


    private void removeCourseInfoFromStudentFile(String studentFileName, String courseToRemove) {
        try {
            FileReader fileReader = new FileReader(studentFileName);
            JSONParser jsonParser = new JSONParser();
            JSONObject studentInFile = (JSONObject) jsonParser.parse(fileReader);

            JSONArray enrolledCoursesArray = (JSONArray) studentInFile.get("enrolledCourses");

            int indexToRemove = -1;
            for (int i = 0; i < enrolledCoursesArray.size(); i++) {
                JSONObject courseInFile = (JSONObject) enrolledCoursesArray.get(i);
                String courseName = (String) courseInFile.get("courseName");
                if (courseName.equalsIgnoreCase(courseToRemove)) {
                    indexToRemove = i;
                    break;
                }
            }
            if (indexToRemove != -1) {
                enrolledCoursesArray.remove(indexToRemove);

                try (FileWriter fileWriter = new FileWriter(studentFileName)) {
                    fileWriter.write(studentInFile.toJSONString());  
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        } catch (IOException | ParseException e) {
            e.printStackTrace();
        }
    }
        private boolean hasEnrollmentFile(Student student) {
            String studentFileName = student.getStudentID() + ".json";
            File studentFile = new File(studentFileName);
            return studentFile.exists();
        }
        
        public void loadTranscriptsFromJson(String filePath) {
            JSONParser parser = new JSONParser();

            try (FileReader reader = new FileReader(filePath)) {
                Object obj = parser.parse(reader);
                JSONArray transcriptArray = (JSONArray) obj;

                for (Object transcriptObj : transcriptArray) {
                    JSONObject transcriptInFile = (JSONObject) transcriptObj;

                    String studentID = (String) transcriptInFile.get("studentID");

                    // Find the student associated with the transcript
                    Student student = findStudentByID(studentID);

                    if (student != null) {
                        Transcript transcript = new Transcript(student.getStudentID());

                        transcript.setCompletedCourses(castToStringList(transcriptInFile.get("completedCourses")));
                        transcript.setFailedCourses(castToStringList(transcriptInFile.get("failedCourses")));
                        transcript.setGPA(((Double) transcriptInFile.get("gpa")).doubleValue());
                        //transcript.setCompletedCredits(((Long) transcriptInFile.get("completedCredits")).intValue());
                        //transcript.setCreditsTaken(((Long) transcriptInFile.get("creditsTaken")).intValue());

                        // Set the transcript for the student
                        student.setTranscript(transcript);
                    }
                }
            } catch (IOException | ParseException e) {
                e.printStackTrace();
            }
        }
        
        private List<String> castToStringList(Object rawList) {
            List<String> stringList = new ArrayList<>();
            if (rawList instanceof List<?>) {
                for (Object item : (List<?>) rawList) {
                    if (item instanceof String) {
                        stringList.add((String) item);
                    }
                    else {
                        // Handle other types or log a warning
                    }
                }
            }
            return stringList;
        }

        Student findStudentByID(String studentID) {
            for (Student student : students) {
                if (student.getStudentID().equals(studentID)) {
                    return student;
                }
            }
            return null;
        }


    


        public static void main(String[] args) {
            CourseRegistrationSystem registrationSystem = new CourseRegistrationSystem();
            registrationSystem.loadStudentsFromJson("students.json");
            registrationSystem.loadCoursesFromJson("courses.json");
            registrationSystem.loadAdvisorsFromJson("advisors.json");
            registrationSystem.loadTranscriptsFromJson("transcripts.json");

            Scanner scanner = new Scanner(System.in);
            boolean continueLogin = true;

            while (continueLogin) {
                System.out.println("How do you wish to enter the system?");
                System.out.println("1. Student");
                System.out.println("2. Advisor");
                System.out.println("0. Exit");

                int choice = scanner.nextInt();

                if (choice == 1) {
                    System.out.print("Enter Student ID: ");
                    String studentID = scanner.next();

                    Student selectedStudent = null;
                    for (Student student : registrationSystem.students) {
                        if (student.getStudentID().equals(studentID)) {
                            selectedStudent = student;
                            break;
                        }
                    }

                    if (selectedStudent != null) {
                        registrationSystem.authenticateAndEnroll(selectedStudent);
                    } else {
                        System.out.println("Student not found. Please try again.");
                    }
                } else if (choice == 2) {
                    System.out.print("Enter Advisor First Name: ");
                    String advisorFirstName = scanner.next().toLowerCase();
                    System.out.print("Enter Advisor Last Name: ");
                    String advisorLastName = scanner.next().toLowerCase();

                    Advisor selectedAdvisor = null;
                    for (Advisor advisor : registrationSystem.advisors) {
                        if (advisor.getFirstName().toLowerCase().equals(advisorFirstName) && advisor.getLastName().toLowerCase().equals(advisorLastName)) {
                            selectedAdvisor = advisor;
                            break;
                        }
                    }

                    if (selectedAdvisor != null) {
                        registrationSystem.authenticateAndEnroll(selectedAdvisor);
                    } else {
                        System.out.println("Advisor not found. Please try again.");
                    }
                }

                 else if (choice == 0) {
                    System.out.println("Exiting...");
                    System.exit(0);
                }
                 else {
                    System.out.println("Invalid choice. Exiting...");
                    continueLogin = false;
                }

                if (continueLogin) {
                    scanner.nextLine();
                    System.out.println("Do you wish to enter the system again? (yes/no)");
                    String again = scanner.nextLine().toLowerCase();
                    if (again.equals("no")) {
                        System.out.println("Exiting...");
                        continueLogin = false;
                    }
                    else if (again.equals("yes")){
                    	// do nothing
                    }
                    else {
                    	System.out.println("Invalid choice. Exiting...");
                    	continueLogin = false;
                    }
                }
            }
        } 
	





public Advisor findAdvisorByName(String firstName, String lastName) {
    for (Advisor advisor : advisors) {
        if (advisor.getFirstName().equalsIgnoreCase(firstName) && advisor.getLastName().equalsIgnoreCase(lastName)) {
            return advisor;
        }
    }
    return null;
}

public Course findCourseByID(String courseID) {
    for (Course course : courses) {
        if (course.getCourseID().equalsIgnoreCase(courseID)) {
            return course;
        }
    }
    return null;
}}

File: Iteration-3/Src/CourseRegistrationSystem.py
import json
import os
from Course import Course
from Student import Student
from Advisor import Advisor
from Transcript import Transcript

class CourseRegistrationSystem:
    def __init__(self):
        self.students = []
        self.courses = []
        self.advisors = []
        self.transcripts = []

    def has_prerequisites(self, student, course):
        prerequisites = course.get_prerequisite()

        if not prerequisites:
            return True

        completed_courses = [c["courseID"] for c in student.get_transcript().get_completed_courses()]
        return any(prerequisite in prerequisites for prerequisite in completed_courses)

    def get_available_courses(self, student):
        available_courses = []

        for course in self.courses:
            if student.get_semester() == course.get_course_semester():
                available_courses.append(course)

        return available_courses

    def get_students(self):
        return self.students

    def get_courses(self):
        return self.courses

    def load_students_from_json(self, file_path):
        try:
            with open(file_path, 'r') as file:
                student_data = json.load(file)
                for student_info in student_data:
                    student = Student(student_info["firstName"], student_info["lastName"],
                                      student_info["studentID"], student_info["password"],
                                      student_info["GPA"], student_info["semester"])
                    self.students.append(student)
        except FileNotFoundError:
            print(f"File not found: {file_path}")

    def load_courses_from_json(self, file_path):
        try:
            with open(file_path, 'r') as file:
                course_data = json.load(file)
                for course_info in course_data:
                    course = Course(course_info["courseID"], course_info["courseName"],
                                    course_info["credit"], None, None,
                                    course_info["prerequisite"], course_info["coursesemester"])
                    self.courses.append(course)
        except FileNotFoundError:
            print(f"File not found: {file_path}")

    def load_advisors_from_json(self, file_path):
        try:
            with open(file_path, 'r') as file:
                advisor_data = json.load(file)
                for advisor_info in advisor_data:
                    advisor = Advisor(advisor_info["firstName"],
                                      advisor_info["lastName"],
                                      advisor_info["password"],
                                      [],
                                      advisor_info["students"])
                    self.advisors.append(advisor)

        except FileNotFoundError:
            print(f"File not found: {file_path}")

    def authenticate_and_enroll(self, person):
        while True:
            email = input("E-mail Address: ").lower()
            password = input("Password: ")

            if isinstance(person, Student):
                current_student = person
                signed_email = current_student.create_email_address().lower()
                signed_password = current_student.get_password()

                if signed_email == email and signed_password == password:
                    self.select_courses(current_student)
                    self.generate_student_files(current_student)
                    again = input("Do you want to exit from system? (yes/no): ").lower()
                    if(again == 'yes'):
                        break
                else:
                    print("Login Failed. Please try again.")
            elif isinstance(person, Advisor):
                current_advisor = person
                signed_email = current_advisor.create_email_address().lower()
                signed_password = current_advisor.get_password()
                if signed_email == email and signed_password == password:
                    print("Advisor" + current_advisor.first_name + current_advisor.last_name + ", login successful.")
                    self.display_enrolled_students_for_advisor(current_advisor)
                    again = input("Do you want to exit from system? (yes/no): ").lower()
                    if(again == 'yes'):
                        break
                else:
                    print("Login Failed. Please try again.")

    def remove_course_info_from_student_file(self, student_file_name, course_info):
        if os.path.exists(student_file_name):
            with open(student_file_name, "r") as file:
                data = json.load(file)
                enrolled_courses = data.get("enrolledCourses")

            if enrolled_courses:
                for course in enrolled_courses:
                    if course["courseID"] == course_info["courseID"]:
                        enrolled_courses.remove(course)

                with open(student_file_name, "w") as file:
                    json.dump(data, file, indent=4)

    def select_courses(self, authenticated_student):
        print(f"Welcome {authenticated_student.first_name} {authenticated_student.last_name}\n")
        available_courses = self.get_available_courses(authenticated_student)

        enrolled_courses = []
        student_file_name = f"{authenticated_student.get_student_id()}.json"
        if os.path.exists(student_file_name):
            with open(student_file_name, "r") as file:
                data = json.load(file)
                enrolled_course_ids = [course['courseID'] for course in data.get("enrolledCourses", [])]
                enrolled_courses = [self.find_course_by_id(course_id) for course_id in enrolled_course_ids]
        available_courses = [course for course in available_courses if course not in enrolled_courses]
        if not available_courses:
            print("There is not any course for your semester.")
            return

        courses_enrolled = 0

        if self.has_enrollment_file(authenticated_student):
            student_file_name = f"{authenticated_student.get_student_id()}.json"
            if os.path.exists(student_file_name):
                with open(student_file_name, "r") as file:
                    data = json.load(file)
                    enrolled_courses = data.get("enrolledCourses")
                    courses_enrolled = len(enrolled_courses) if enrolled_courses else 0

        while courses_enrolled < 5:
            if not available_courses:
                print("There is no available course.")
                break

            print("Available Courses:")

            for i, course in enumerate(available_courses, start=1):
                print(f"{i}. {course.get_course_name()}")

            print("To choose courses, enter their numbers line by line or press \"0\" to exit from selection system:")
            user_input = input()

            if user_input == "0":
                break
            selected_course_numbers = user_input.split(",")

            for course_number in selected_course_numbers:
                try:
                    choice = int(course_number)

                    if 1 <= choice <= len(available_courses):
                        selected_course = available_courses[choice - 1]

                        if self.has_prerequisites(authenticated_student, selected_course):

                            authenticated_student.add_course(selected_course)
                            print(f"You enrolled {selected_course.get_course_id()} - {selected_course.get_course_name()}.")
                            available_courses.remove(selected_course)
                            courses_enrolled += 1
                        else:
                            print(f"You has not completed prerequisite of {selected_course.get_course_id()} yet.")
                    else:
                        print(f"Invalid Choice: {course_number}.")
                except ValueError:
                    print(f"Invalid choice: {course_number}.")

    def generate_student_files(self, student):
        student_file_name = f"{student.get_student_id()}.json"
        if os.path.exists(student_file_name):
            with open(student_file_name, "r") as file:
                student_data = json.load(file)
                # Assume that 'enrolledCourses' is a list of course dictionaries
                existing_courses = {course["courseID"]: course for course in student_data.get("enrolledCourses", [])}
        else:
            student_data = {
                "studentID": student.get_student_id(),
                "firstName": student.get_first_name(),
                "lastName": student.get_last_name(),
                "semester": student.get_semester(),
                "GPA": student.get_gpa(),
                "enrolledCourses": []
            }
            existing_courses = {}

        # Add new courses to the existing ones, avoiding duplicates
        for course in student.get_courses():
            if course.get_course_id() not in existing_courses:
                course_data = {
                    "courseID": course.get_course_id(),
                    "courseName": course.get_course_name(),
                    "credit": course.get_credit(),
                    "prerequisite": course.get_prerequisite(),
                    "coursesemester": course.get_course_semester()
                }
                student_data["enrolledCourses"].append(course_data)

        # Save the updated data back to the JSON file
        with open(student_file_name, "w") as file:
            json.dump(student_data, file, indent=4)

    def display_enrolled_students_for_advisor(self, advisor):
        print(f"Welcome {advisor.get_first_name()} {advisor.get_last_name()}, your students: ")

        while True:
            for student in advisor.get_students():
                print(f"{student['firstName']} {student['lastName']} - {student['studentID']}")

            student_id = input("Enter a student ID or type 'exit' to exit from system: ")

            if student_id.lower() == 'exit':
                break
            selected_student = self.find_student_by_id(student_id)
            if selected_student:
                self.approve_courses_for_student(selected_student)
            else:
                print("Student can not be found.")

    def approve_courses_for_student(self, student):
        student_file_name = f"{student.get_student_id()}.json"
        if os.path.exists(student_file_name):
            with open(student_file_name, "r") as file:
                data = json.load(file)
                enrolled_courses = data.get("enrolledCourses")
                for course in enrolled_courses:
                    print(f"To approve/deny the Course: {course['courseID']} - {course['courseName']} (yes/no):")
                    approval = input().lower()
                    if approval == 'yes':
                        print(f"{course['courseID']} - {course['courseName']} onaylandı.")
                    else:
                        print(f"{course['courseID']} - {course['courseName']} reddedildi.")
                        self.remove_course_info_from_student_file(student_file_name, course)
                print(f"Courses approving/denying process is completed for this student.")
        else:
            print("This student has not enrolled any courses yet.")

    def has_enrollment_file(self, student):
        student_file_name = f"{student.get_student_id()}.json"
        return os.path.exists(student_file_name)

    def load_transcripts_from_json(self, file_path):
        try:
            with open(file_path, 'r') as file:
                transcript_data = json.load(file)
                for transcript_info in transcript_data:
                    student = self.find_student_by_id(transcript_info["studentID"])
                    if student:
                        transcript = Transcript(student.get_student_id())
                        transcript.set_completed_courses(transcript_info["completedCourses"])
                        transcript.set_failed_courses(transcript_info["failedCourses"])
                        transcript.set_gpa(transcript_info["gpa"])
                        student.set_transcript(transcript)

        except FileNotFoundError:
            print(f"File not found: {file_path}")
        except json.JSONDecodeError as e:
            print(f"Error occurred during the parsing of JSON file: {e}")

    def find_advisor_by_name(self, first_name, last_name):
        for advisor in self.advisors:
            if advisor.get_first_name().lower() == first_name.lower() and advisor.get_last_name().lower() == last_name.lower():
                return advisor
        return None

    def find_course_by_id(self, course_id):
        for course in self.courses:
            if course.get_course_id().lower() == course_id.lower():
                return course
        return None

    def find_student_by_id(self, student_id):
        for student in self.students:
            if student.get_student_id().lower() == student_id.lower():
                return student
        return None
if __name__ == "__main__":
    registration_system = CourseRegistrationSystem()
    registration_system.load_students_from_json("students.json")
    registration_system.load_courses_from_json("courses.json")
    registration_system.load_advisors_from_json("advisors.json")
    registration_system.load_transcripts_from_json("transcripts.json")

    continue_login = True

    while continue_login:
        print("How do you wish to enter the system?")
        print("1. Student")
        print("2. Advisor")
        print("0. Exit")

        choice = input()

        if choice == '1':
            student_id = input("Enter Student ID: ")


            selected_student = None
            for student in registration_system.students:
                if student.get_student_id() == student_id:
                    selected_student = student
                    break

            if selected_student is not None:
                registration_system.authenticate_and_enroll(selected_student)
            else:
                print("Student not found. Please try again.")
        elif choice == '2':

            advisor_first_name = input("Enter Advisor First Name: ").lower()
            advisor_last_name = input("Enter Advisor Last Name: ").lower()
            selected_advisor = None
            for advisor in registration_system.advisors:
                if advisor.get_first_name().lower() == advisor_first_name and advisor.get_last_name().lower() == advisor_last_name:
                    selected_advisor = advisor
                    break

            if selected_advisor is not None:
                registration_system.authenticate_and_enroll(selected_advisor)
            else:
                print("Advisor not found. Please try again.")
        elif choice == '0':
            print("Exiting...")
            continue_login = False
        else:
            print("Invalid choice. Exiting...")

        if continue_login:
            print("Do you wish to enter the system again? (yes/no)")
            again = input().lower()
            if again == "no":
                print("Exiting...")
                continue_login = False
            elif again != "yes":
                print("Invalid choice. Exiting...")
                continue_login = False
